--- a/dwl.c
+++ b/dwl.c
@@ -341,6 +341,7 @@
 static void tag(const Arg *arg);
 static void tagmon(const Arg *arg);
 static void tile(Monitor *m);
+static void bstack(Monitor *m);
 static void togglefloating(const Arg *arg);
 static void togglefullscreen(const Arg *arg);
 static void toggletag(const Arg *arg);
@@ -2816,6 +2817,11 @@
 	if (n == 0)
 		return;
 
+	if (m->w.height > m->w.width) {
+		bstack(m);
+		return;
+	}
+
 	if (n > m->nmaster)
 		mw = m->nmaster ? (int)roundf(m->w.width * m->mfact) : 0;
 	else
@@ -2835,6 +2841,40 @@
 		}
 		i++;
 	}
+}
+
+void
+bstack(Monitor *m)
+{
+	unsigned int mh, mx, tx;
+	int i, n = 0;
+	Client *c;
+
+	wl_list_for_each(c, &clients, link)
+		if (VISIBLEON(c, m) && !c->isfloating && !c->isfullscreen)
+			n++;
+	if (n == 0)
+		return;
+
+	if (n > m->nmaster)
+		mh = m->nmaster ? (int)roundf(m->w.height * m->mfact) : 0;
+	else
+		mh = m->w.height;
+	i = mx = tx = 0;
+	wl_list_for_each(c, &clients, link) {
+		if (!VISIBLEON(c, m) || c->isfloating || c->isfullscreen)
+			continue;
+		if (i < m->nmaster) {
+			resize(c, (struct wlr_box){.x = m->w.x + mx, .y = m->w.y,
+				.width = (m->w.width - mx) / (MIN(n, m->nmaster) - i), .height = mh}, 0);
+			mx += c->geom.width;
+		} else {
+			resize(c, (struct wlr_box){.x = m->w.x + tx, .y = m->w.y + mh,
+				.width = (m->w.width - tx) / (n - i), .height = m->w.height - mh}, 0);
+			tx += c->geom.width;
+		}
+		i++;
+	}
 }
 
 void
